### PR TITLE
BUG: cluster.vq: ensure `_kmeans` returns distortion corresponding to final codebook

### LINKED ISSUE #24069: BUG: `scipy.cluster.vq.kmeans` returns a distortion value from a different iteration than the final centroids
### Describe your issue.

`kmeans` should return `(codebook, distortion)` where `distortion` is “the mean (non-squared) Euclidean distance between the observations passed and the centroids generated.” However, when the algorithm runs multiple random initializations, the function sometimes returns the centroids from one iteration and the average distance from another. The result is a distortion value that does not match `np.mean(vq(obs, codebook)[1])`.

### Reproducing Code Example

```python
import numpy as np
from scipy.cluster.vq import kmeans, vq, whiten

rows = [[8.0, 1e-5, 1e-5, 1e-5]]          # one outlier
rows += [[1e-5, 1e-5, 1e-5, 1e-5]] * 51   # many identical rows
rows += [[0.0, 1e-5, 1e-5, 1e-5]]         # another distinct row

obs = np.array(rows, dtype=np.float64)
wh = whiten(obs)

# Run kmeans several times to see the mismatch
for i in range(10):
    codebook, distortion = kmeans(wh, 3)   # default iter=20
    _, dists = vq(wh, codebook)
    mean_dist = float(np.mean(dists))
    if abs(mean_dist - distortion) > 1e-6:
        print(f"mismatch on run {i}: distortion={distortion}, mean={mean_dist}")
        break
else:
    print("no mismatch observed")
```

### Error message

```shell
mismatch on run 3: distortion=0.1386840596118993, mean=1.8447368768250103
```

### SciPy/NumPy/Python version and system information

```shell
1.16.3 2.3.4 sys.version_info(major=3, minor=12, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.12.11/lib/python3.12/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.12.11/lib/python3.12/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.29.dev DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.29.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /opt/_internal/cpython-3.12.11/lib/python3.12/site-packages/scipy_openblas32/include
    lib directory: /opt/_internal/cpython-3.12.11/lib/python3.12/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.29.dev DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /project
    version: 0.3.29.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 3.0.1
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.1.6
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/build-env-99e64tom/lib/python3.12/site-packages/pythran
    version: 0.18.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /tmp/build-env-99e64tom/bin/python
  version: '3.12'
```

### PR DESCRIPTION
**Pull request by BarnikRB**
#### Reference issue
https://github.com/scipy/scipy/issues/24069

#### What does this implement/fix?
The internal _kmeans routine returned a distortion value computed for a previous iteration rather than for the final codebook (after centroid updates and empty-cluster removal). This led to cases where distortion did not match mean(vq(obs, codebook)[1]).

This change recomputes the distortion with respect to the final codebook before returning, ensuring the reported distortion is consistent with the returned centroids.

### CODE REVIEW

**Review by lucascolley — COMMENTED** [review]
Thanks @BarnikRB, this looks good to me in principle. Please can you check if this behaviour matches https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html ? I think it is the `inertia_` attribute that is relevant.

**Comment by BarnikRB** [comment]
Hello @lucascolley ,

I just checked the kmeans code from the sklearn repo. In scikit-learn, the clustering objective is always computed using the final cluster centers. In  `_kmeans_single_lloyd` ([line 630](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/_kmeans.py#L630)), after convergence the E-step is rerun (if needed) so that labels match the final centers, and then inertia is computed using the final labels and centers:

```
if not strict_convergence:
    lloyd_iter(..., update_centers=False)

inertia = _inertia(X, sample_weight, centers, labels, n_threads)

```

So, scipy kmeans also should calculate the distortion using the final centroids.

File: scipy/cluster/vq.py [hunk 1]
```diff
@@ -325,8 +325,9 @@ def _kmeans(obs, guess, thresh=1e-5, xp=None):
         code_book = code_book[has_members]
         code_book = xp.asarray(code_book)
         diff = xp.abs(prev_avg_dists[0] - prev_avg_dists[1])
-
-    return code_book, prev_avg_dists[1]
+    _,final_distortions = vq(obs, code_book, check_finite=False)
```
Comment (lucascolley) [inline scipy/cluster/vq.py]:
```suggestion

    _, final_distortions = vq(obs, code_book, check_finite=False)
```

**Review by lucascolley — APPROVED** [review]
thanks @barnikRB!
